src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -111,5 +111,52 @@ private static IEnumerable<ResourceDefinition> GetResourceDefinitions(SemanticMo
                 yield return new ResourceDefinition(resource.Name, scopeSymbol, resourceType.TypeReference.FullyQualifiedType, namePropertyValue);
             }
         }
+
+        public static void DetectIncorrectlyFormattedNames(SemanticModel semanticModel, IDiagnosticWriter diagnosticWriter)
+        {
+            foreach (var resource in semanticModel.Root.GetAllResourceDeclarations())
+            {
+                if (semanticModel.GetTypeInfo(resource.DeclaringSyntax) is not ResourceType resourceType)
+                {
+                    continue;
+                }
+
+                var resourceName = resource.SafeGetBodyPropertyValue(LanguageConstants.ResourceNamePropertyName);
+                if (resourceName is not StringSyntax resourceNameString)
+                {
+                    // not easy to do analysis if it's not a string!
+                    continue;
+                }
+
+                var ancestors = semanticModel.ResourceAncestors.GetAncestors(resource);
+                if (ancestors.Any())
+                {
+                    // try to detect cases where someone has applied top-level resource declaration naming to a nested/parent resource
+                    // e.g. '{parent.name}/child' or 'parent/child'
+                    if (resourceNameString.SegmentValues.Any(v => v.IndexOf('/') > -1))
+                    {
+                        // TODO fix this
+                        diagnosticWriter.Write(resourceName, x => x.NestedChildResourceNameContainsQualifiers());
+                    }
+                }
+                else
+                {
+                    if (resourceNameString.TryGetLiteralValue() is {} typeString)
+                    {
+                        // try to detect cases where someone has applied nested/parent resource declaration naming to a top-level resource - e.g. 'child'.
+                        // this unfortunately limits us to only validating uninterpolated strings, as we can't do analysis on the number of '/' characters
+                        // being pulled in from variables.
+                        
+                        var slashCount = typeString.Count(x => x == '/');
+                        var expectedSlashCount = resourceType.TypeReference.Types.Length - 1;
+                        if (slashCount != expectedSlashCount)
+                        {
+                            // TODO fix this
+                            diagnosticWriter.Write(resourceName, x => x.TopLevelChildResourceNameMissingQualifiers(expectedSlashCount));
+                        }                        
```

Comment:
I wonder if we could also check if there's at least one character before and after a slash. Verifying the number of slashes only doesn't block names that start/end with '/' and names containing adjacent '/'s.
```
resource vmExt 'Microsoft.Compute/virtualMachines/extensions@2020-06-01' = {
  name: 'myExt/'
  location: 'eastus'
}


resource attachment 'Microsoft.ApiManagement/service/apis/issues/attachments@2019-01-01' = {
    name: '///'
}
```

Reply:
Btw just saw that there's a TODO, so not sure you are already planning on doing this 😀

Reply:
The reason I added this logic was to prevent people tripping up on the difference between 'no parent property' and 'parent property' - e.g. making sure we displayed an error for anyone attempting to convert between the following syntaxes without adjusting the name to account:
```bicep
resource vmExt 'Microsoft.Compute/virtualMachines/extensions@2020-06-01' = {
  name: '${vm.name}/myExt'
}
```
and 
```bicep
resource vmExt 'Microsoft.Compute/virtualMachines/extensions@2020-06-01' = {
  parent: vm
  name: 'myExt'
}
```

We could definitely go further with the naming validation, but it feels like a separate issue.

Reply:
Yeah that makes sense!